### ISSUE TITLE
Build failure

### ISSUE THREAD

**Issue by rnestler** [woboq/qmetaobject-rs#12]
I tried to build the latest master version on my machine and got the following errors:

<details>
<summary>Build output. Click to expand</summary>  <p>

```
   Compiling qmetaobject v0.0.1 (file:///home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject)    
error: failed to run custom build command for `qmetaobject v0.0.1 (file:///home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject)`
process didn't exit successfully: `/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/target/debug/build/qmetaobject-4487a1e747466dad/build-script-build` (exit code: 101)
--- stdout
TARGET = Some("x86_64-unknown-linux-gnu")
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-linux-gnu")
CXX_x86_64-unknown-linux-gnu = None
CXX_x86_64_unknown_linux_gnu = None
HOST_CXX = None
CXX = None
CXXFLAGS_x86_64-unknown-linux-gnu = None
CXXFLAGS_x86_64_unknown_linux_gnu = None
HOST_CXXFLAGS = None
CXXFLAGS = None
DEBUG = Some("true")
CXX_x86_64-unknown-linux-gnu = None
CXX_x86_64_unknown_linux_gnu = None
HOST_CXX = None
CXX = None
CXXFLAGS_x86_64-unknown-linux-gnu = None
CXXFLAGS_x86_64_unknown_linux_gnu = None
HOST_CXXFLAGS = None
CXXFLAGS = None
running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject" "-I" "/usr/include/x86_64-linux-gnu/qt5" "-Wall" "-Wextra" "-std=c++11" "-o" "/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/target/debug/build/qmetaobject-8866b0bf71f118a5/out/rust_cpp/cpp_closures.o" "-c" "/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/target/debug/build/qmetaobject-8866b0bf71f118a5/out/rust_cpp/cpp_closures.cpp"
cargo:warning=src/lib.rs: In function ‘void __cpp_closure_4690954296995772382(FnBoxWrapper&, const QPointer<QThread>&)’:
cargo:warning=src/lib.rs:561:94: error: no matching function for call to ‘QMetaObject::invokeMethod(QObject*&, std::remove_reference<FnBoxWrapper&>::type, Qt::ConnectionType)’
cargo:warning=                 QMetaObject::invokeMethod(reciever, std::move(func_raw), Qt::QueuedConnection); // does not allow move-only
cargo:warning=                                                                                              ^
cargo:warning=In file included from /usr/include/x86_64-linux-gnu/qt5/QtCore/qobject.h:46:0,
cargo:warning=                 from /usr/include/x86_64-linux-gnu/qt5/QtCore/QObject:1,
cargo:warning=                 from /home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/qmetaobject_rust.hpp:20,
cargo:warning=                 from src/lib.rs:51:
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:406:17: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                 ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:406:17: note:   candidate expects 14 arguments, 3 provided
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:420:24: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static inline bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                        ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:420:24: note:   no known conversion for argument 2 from ‘std::remove_reference<FnBoxWrapper&>::type {aka FnBoxWrapper}’ to ‘const char*’
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:437:24: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, Qt::ConnectionType, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static inline bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                        ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:437:24: note:   no known conversion for argument 2 from ‘std::remove_reference<FnBoxWrapper&>::type {aka FnBoxWrapper}’ to ‘const char*’
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:454:24: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static inline bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                        ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:454:24: note:   no known conversion for argument 2 from ‘std::remove_reference<FnBoxWrapper&>::type {aka FnBoxWrapper}’ to ‘const char*’
cargo:warning=src/lib.rs:564:90: error: no matching function for call to ‘QMetaObject::invokeMethod(QGuiApplication*, std::remove_reference<FnBoxWrapper&>::type, Qt::ConnectionType)’
cargo:warning=                 QMetaObject::invokeMethod(qApp, std::move(func_raw), Qt::QueuedConnection);
cargo:warning=                                                                                          ^
cargo:warning=In file included from /usr/include/x86_64-linux-gnu/qt5/QtCore/qobject.h:46:0,
cargo:warning=                 from /usr/include/x86_64-linux-gnu/qt5/QtCore/QObject:1,
cargo:warning=                 from /home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/qmetaobject_rust.hpp:20,
cargo:warning=                 from src/lib.rs:51:
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:406:17: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                 ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:406:17: note:   candidate expects 14 arguments, 3 provided
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:420:24: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static inline bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                        ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:420:24: note:   no known conversion for argument 2 from ‘std::remove_reference<FnBoxWrapper&>::type {aka FnBoxWrapper}’ to ‘const char*’
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:437:24: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, Qt::ConnectionType, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static inline bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                        ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:437:24: note:   no known conversion for argument 2 from ‘std::remove_reference<FnBoxWrapper&>::type {aka FnBoxWrapper}’ to ‘const char*’
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:454:24: note: candidate: static bool QMetaObject::invokeMethod(QObject*, const char*, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument)
cargo:warning=     static inline bool invokeMethod(QObject *obj, const char *member,
cargo:warning=                        ^~~~~~~~~~~~
cargo:warning=/usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs.h:454:24: note:   no known conversion for argument 2 from ‘std::remove_reference<FnBoxWrapper&>::type {aka FnBoxWrapper}’ to ‘const char*’
exit code: 1

--- stderr
thread 'main' panicked at '

Internal error occurred: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject" "-I" "/usr/include/x86_64-linux-gnu/qt5" "-Wall" "-Wextra" "-std=c++11" "-o" "/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/target/debug/build/qmetaobject-8866b0bf71f118a5/out/rust_cpp/cpp_closures.o" "-c" "/home/rnestler/projects/mso-sw/lin/qmetaobject-rs/qmetaobject/target/debug/build/qmetaobject-8866b0bf71f118a5/out/rust_cpp/cpp_closures.cpp" with args "c++" did not execute successfully (status code exit code: 1).
```
</p></details>

Is this dependent on the Qt version? Also which commit corresponds to the published crate on crates.io (which builds for me)? I didn't find any tag in the repository.

**Comment by ogoffart** [comment]
The API was added with Qt 5.10.
I made a commit to work around it.  Maybe this fixes your problem.
I'm only testeing with Qt 5.10 in the CI, and with Qt 5.11 locally

**Comment by rnestler** [comment]
I'm using Qt 5.9 on Ubuntu 18.04.

**Comment by ogoffart** [comment]
Did the last Commit fix the build for you?

**Comment by rnestler** [comment]
> Did the last Commit fix the build for you?

Yes it now builds on Qt 5.9 :smiley: thanks!

**Comment by rnestler** [comment]
I'm now testing on the Raspberry Pi, but I can open a new issue when I run into any trouble there.

**Comment by ogoffart** [comment]
I added Qt 5.9 and Qt 5.8 in travis

Qt 5.7 does not work because some scenegraph API that I need was only added in Qt 5.8
It would be impossible without it, so Qt 5.7 can't be supported, unless we disable the scenegraph integration when using Qt 5.7.

I considered also supporting Qt 5.6 since it is a LTS, this would (in addition to the scene graph api- just require using a replacement for QMetaObject::inherits which was added in Qt 5.7

So you probably will want to recompile Qt for the rasbery pi. 
And in general, I want to support Qt 5.9 since it is a LTS, but older version probably are not worth it.

I'm closing this issue since the build issue are fixed and Qt 5.9 works.
